docs/proposals/rescheduler.md

```diff
@@ -0,0 +1,130 @@
+[here](http://releases.k8s.io/release-1.0/docs/proposals/rescheduler.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Rescheduler design space
+
+@davidopp, @erictune, @briangrant
+
+July 2015
+
+A rescheduler is an agent that proactively causes currently-running
+Pods to be moved, so as to optimize some objective function for
+goodness of the layout of Pods in the cluster. (The objective function
+doesn't have to be expressed mathematically; it may just be a
+collection of ad-hoc rules, but in principle there is an objective
+function. Implicitly an objective function is described by the
+scheduler's predicate and priority functions.) It might be triggered
+to run every N minutes, or whenever some event happens that is known
+to make the objective function worse (for example, whenever a Pod goes
+PENDING for a long time.)
+
+A rescheduler is useful because without a rescheduler, scheduling
+decisions are only made at the time Pods are created. But as the
+cluster layout changes over time, free "holes" are often produced that
+were not available when a Pod was initially scheduled. These holes are
+produced by run-to-completion Pods terminating, empty nodes being
+added by a node auto-scaler, etc. Moving already-running Pods into
+these holes may lead to a better cluster layout. A rescheduler might
+not just exploit existing holes, but also create holes by evicting
+Pods (assuming it knows they can reschedule elsewhere), as in free
+space defragmentation.
+
+[Although alluded to above, it's worth emphasizing that rescheduling
+is the only way to make use of new nodes added by a cluster
+auto-scaler (unless Pods were already PENDING; but even then, it's
+likely advantageous to put more than just the previously PENDING Pods
+on the new nodes.)]
+
+Because rescheduling is disruptive--it causes one or more
+already-running Pods to die when they otherwise wouldn't--a key
+constraint on rescheduling is that it must be done subject to
+disruption SLOs. There are a number of ways to specify these SLOs--a
+global rate limit across all Pods, a rate limit across a set of Pods
+defined by some particular label selector, a maximum number of Pods
+that can be down at any one time among a set defined by some
+particular label selector, etc. These policies are presumably part of
+the Rescheduler's configuration.
+
+There are a lot of design possibilities for a rescheduler. To explain
+them, it's easiest to start with the description of a baseline
+rescheduler, and then describe possible modifications. The Baseline
+rescheduler
+* only kicks in when there are one or more PENDING Pods for some period of time; its objective function is binary: completely happy if there are no PENDING Pods, and completely unhappy if there are PENDING Pods; it does not try to optimize for any other aspect of cluster layout
```

Comment:
Yes. I considered writing the doc in a more traditional style where I lay out the axes of the design space and explain the different points in the space. However, after struggling with that a bit, I realized it was easier to start with one point in the design space (the Baseline rescheduler) and then describe variations on it. But either way, this is just talking about different design possibilities. Let me know if I am not fully understanding your point.